### ISSUE TITLE
Compiling scenario data for ReMaP plug-in

### ISSUE THREAD

**Issue by shanshanhsieh** [architecture-building-systems/CityEnergyAnalyst#2902]
Prepare a list of mock data that stores the future scenario data from VILLE.
here is the list of data:
1. 18 scenarios folders for building uses (3 district archetypes * 2 dates * 3*urban development scenarios)
2. updated files that contain new construction properties and building systems 

besides, mapping tables are required to store the corresponding properties in different years (2040/2060)

**Comment by nasty-bos** [comment]
I have prepared CH folder with mock data that follows the logic below: 

1. For urban development scenarios, we create an archetype folder for each scenario containing updated building use types. Find 18 scenario folders in the `CH\archetypes` attached. Folders are created with a code DISTRICT_YEAR_URBAN-DEVELOPMENT (e.g. RRL_2040_BAU). Dictionary of codes below: 

Urban development scenarios: 
- 	Business-as-usual (BAU)
- 	Urban decline (DCL)
- 	Balanced (BLN)

District archetypes:

- 	Urban (URB)
- 	Sub-urban (SURB)
- 	Rural (RRL)

2. For retrofit and systems scenarios, we create new elements in the existing files (new component, new standard, new assembly etc.). This means, the files `CONSTRUCTION_STANDARD.xlsx`, `ENVELOPE.xlsx`, `SUPPLY.xlsx`, `HVAC.xlsx` will contain updated elements as new rows.  We use existing code naming rules. For now, I added one new element to each table. 

3. Mapping tables will allow to set the correct component combinations for studied scenarios. Find mapping tables for each component of ENVELOPE, SUPPLY, HVAC attached. Each file contains mapping for elements present in the initial` ENVELOPE.xlsx`, `SUPPLY.xlsx`,` HVAC.xlsx` files. For example, window type WINDOW_AS1 from status quo scenario will be replaced by a new window type WINDOW_AS11 which properties will be defined in a new row of `ENVELOPE.xlsx`. This change will depend on the district, building use, year and retrofit scenario (business-as-usual BAU or new energy politics NEP):
![image](https://user-images.githubusercontent.com/37011711/103804564-9eff3d80-5052-11eb-8ca5-8ad0be1defba.png)

@daren-thomas, @shanshanhsieh  looking forward to your feedback. 

[mapping_HVAC.xlsx](https://github.com/architecture-building-systems/CityEnergyAnalyst/files/5777464/mapping_HVAC.xlsx)
[mapping_SUPPLY.xlsx](https://github.com/architecture-building-systems/CityEnergyAnalyst/files/5777465/mapping_SUPPLY.xlsx)
[mapping_ENVELOPE.xlsx](https://github.com/architecture-building-systems/CityEnergyAnalyst/files/5777466/mapping_ENVELOPE.xlsx)
[CH_ReMaP.zip](https://github.com/architecture-building-systems/CityEnergyAnalyst/files/5777470/CH_ReMaP.zip)

**Comment by nasty-bos** [comment]
An alternative to point 3 would be to create a mapping for CONSTRUCTION_STANDARD only, in order to reduce the amount of mapping rules. Instead of going by each element (window, floor etc.), we could define a bunch of new construction standards in `CONSTRUCTION_STANDARD.xlsx` and use one mapping file (see attached). 
[mapping_CONSTRUCTION_STANDARD.xlsx](https://github.com/architecture-building-systems/CityEnergyAnalyst/files/5777623/mapping_CONSTRUCTION_STANDARD.xlsx)

**Comment by daren-thomas** [comment]
@nasty-bos uploaded mock data to be reviewed this week with @daren-thomas (Friday, outlook invitation sent)

**Comment by nasty-bos** [comment]
After discussion with @daren-thomas, we decided to use mapping for CONSTRUCTION_STANDARD. See updated mapping table attached to avoid MultiIndex. There are two ways to do that, use whichever you prefer. 

[mapping_CONSTRUCTION_STANDARD_v1.xlsx](https://github.com/architecture-building-systems/CityEnergyAnalyst/files/5808942/mapping_CONSTRUCTION_STANDARD_v1.xlsx)
[mapping_CONSTRUCTION_STANDARD_v2.xlsx](https://github.com/architecture-building-systems/CityEnergyAnalyst/files/5808943/mapping_CONSTRUCTION_STANDARD_v2.xlsx)

**Comment by nasty-bos** [comment]
ReMaP plug-in should have additional feature that updates use types using a mapping table ([mapping_USE_TYPE.xlsx](https://github.com/architecture-building-systems/CityEnergyAnalyst/files/5821862/mapping_USE_TYPE.xlsx)). This will allow to update building uses depending on the urban development scenario, both for mixed-use and single-use buildings. There are 3 situations to be considered: 1) building use remains the same 2) building is converted to another use type 3) building use ratio in mixed-use buildings changes. All of them can be covered by creating a new row in the mapping table. 
Mapping table is based on the naming rules from ` typology.dbf `. It is structured to depict status_quo_building_use (up to 3 components) and updated_building_use for 3 components with corresponding ratios, depending on the urban development scenario (e.g. RRL_2040_BLN, SURB_2040_DCL etc.)
Mapping of use types should be done prior to construction standards mapping. 

Optional feature to treat mixed-use buildings: it reads ratios of mixed-use buildings from ` typology.dbf ` of the status-quo scenario and compares them to a mapping table. For example, if initial ratio of certain use is 0.997, it will look for a rule in the mapping table in the range of 0.9-1 (because we will not be able to create mapping rules manually with a step of 0.001 :D). Then, it allows to assign new ratios based on a rule system. For example, `IF STATUS_QUO_1ST_USE = "OFFICE" AND STATUS_QUO_1ST_USE_R<0.5 THEN RRL_2040_BLN_1ST_USE = "SINGLE_RES" AND STATUS_QUO_1ST_USE_R=1`. 

@shanshanhsieh @daren-thomas If there are any ways to make use types mapping smarter, let's discuss! Optional feature would definitely need further thoughts.

**Comment by nasty-bos** [comment]
Another feature that plug-in should have is related to the density change in a district depending on the urban development scenarios. We need to represent 4 possible ways to change density in a district: 

Geometry-related, meaning update of `zone.shp` 
1. Total built area changes by increasing/decreasing footprint area of the buildings.
2. Total built area changes by adding/removing floors in the buildings. 
3. New building is being built or building is removed. 

Other 
4. Change of occupant density in `internal-loads.dbf`

**Comment by shanshanhsieh** [comment]
taking advises from @daren-thomas, and applying the principle of specifying clear input and output parameters in each mapper file, i am proposing these two formats for use type mapping:
1. one tab for each **case_year_scenario** (e.g., RRL_2040_BAU): the three uses and ratios are specified according to the use types in the base year. we will have 18 tabs, each with 9 columns, in this case.
2. one tab for each **case_scenario** (e.g., RRL_BAU): the three uses and ratios in 2040 and 2060 are specified according to the use types in the base year. we will have 9 tabs, each with 15 tabs, in this case.

[https://app.zenhub.com/files/49491341/6121cccb-9b11-4a74-b28d-fcee44c1c5c2/download](https://app.zenhub.com/files/49491341/6121cccb-9b11-4a74-b28d-fcee44c1c5c2/download)

the 1st solution is mirroring the folder structure of the CH_remap database
the 2nd solution is more concise

what do you think, @daren-thomas ?

**Comment by daren-thomas** [comment]
`cea remapper --scenario xyz --urban-development-scenario BAU --year 2040 --district-archetype RRL`

- `--scenario`: name of a CEA scenario, requirement that the construction standards in `typology.dbf` are all found in the mapping tables
- `--urban-development-scenario` can be one of [BAU, DCL, BLN]
- `--year` can be one of [2040, 2060]
- `--district-archetype` can be one of [URB, SURB, RRL]

- look up construction standard in `typology.dbf` and apply mappers
- copy over the appropriate archetypes based on urban-development-scenario, year and district-archetype, including use-types folder

- use-type mapping... (update 1st-use etc. incl. ratios based on mapping table)

**Comment by daren-thomas** [comment]
three scripts:

- `cea remap-use-types` (hidden in gui?)
- `cea remap-construction-standards` (hidden in gui?)
- `cea remap` (calls `cea remap-use-types`, then `cea reamap-construction-standards`)

```
(MULTI_RES, 0.7, SINGLE_RES, 0.3,  ANY) => (MULTI_RES, 0.5, SINGLE_RES, 0.2, *ANY, *REST*)
(MULTI_RES, ANY, ANY, ANY, ANY, ANY) => (SINGLE_RES, 0.7, MATCH, MATCH, MATCH, MATCH)
(ANY, ANY, ANY, ANY, any, any) => (MULTI_RES, 1.0, NONE, 0.0, NONE, 0.0) # default
(ANY, ANY, ANY, ANY, any, any) => (COPY,  COPY, COPY, MATCH, MATCH, MATCH)
PATTERN => OUTPUT VALUES...
```

Format for `mapping_USE_TYPE.xlsx`:

- 12 columns, 6 on left (pattern), 6 on right (result)
- ratio columns in pattern specify minimum value (>=)
- ANY matches any value (for USE_TYPE and USE_TYPE_R)
- in result, "COPY" copies value from left hand side
- last ratio for result is computed (1 - 1ST_USE_R - 2ND_USE_R) - ignores last column in result...

Each worksheet is named (`RRL_2040_BAU`), [district-archetype, year, urban-development-scenario].

**Comment by shanshanhsieh** [comment]
here is the updated use type mapping table which contains **18 tabs** and **6 columns** each according to our discussion : )
[mapping_USE_TYPE_v3.xlsx](https://github.com/architecture-building-systems/CityEnergyAnalyst/files/5848201/mapping_USE_TYPE_v3.xlsx)

**Comment by shanshanhsieh** [comment]
the resulting data structure has been implemented in #2903, therefore, this issue is closed.
